### ISSUE TITLE
Additional test cases for csv setCharToEscapeQuoteEscaping 

### ISSUE THREAD

**Issue by ep1804** [uniVocity/univocity-parsers#143]
Based on the documentation [here](https://github.com/uniVocity/univocity-parsers/blob/e34a149077147de7cafcafbbaf8a4310e6297584/src/main/java/com/univocity/parsers/csv/CsvFormat.java#L160-L176), many users will set `charToEscapeQuoteEscaping` to `\`. 

However this setting does not work well sometimes. Here are new test cases:
https://github.com/apache/spark/pull/17177#issuecomment-284607257

IMHO, 

* Default setting should be `charToEscapeQuoteEscaping = quoteChar ` 
  - provided that `quoteChar != quoteEscapeChar`
* Documentation should be updated

How do you think about this?

**Comment by jbax** [comment]
This is automatically determined in the `CsvFormat` class:

public final char getCharToEscapeQuoteEscaping() {
		if (charToEscapeQuoteEscaping == null) { //not provided by the user
			if (quote == quoteEscape) {
				return '\0'; //not required
			} else {
				return quoteEscape;
			}
		}
		return charToEscapeQuoteEscaping;
	}

So:
* if quote and escape are `"`, there no escaping of the escape character is required
* if the escape character is not the same as the quote character, then two occurrences of the escape character are considered to represent a single escape character. E.g. `\\` represents `\`.

Users will hardly ever need to fiddle with the `charToEscapeQuoteEscaping`. I'll take a look at your tests later so verify if there is any issue. Thanks for reporting this.

**Comment by jbax** [comment]
It seems all your test cases involve parsing unescaped quotes (i.e. **broken CSV**) so I believe you are getting confused by what the parser spits out as it tries to salvage the data by default instead of throwing an error at you.

No matter what combination of escape escape you are setting for these tests, all inputs are invalid CSV, especially if you change the escape escape to be different than the escape character.

In the following inputs I replaced your double quotes by single quotes (`'`), and the escape character a as pipe (`|`) to make them easier to visualize as Java Strings)

Here is my test case (that currently just prints to the sysout):

```
	@DataProvider
	Object[][] data() {
		return new Object[][]{
				{"'''AA'BB'''", '\''},      // 1 quote char (OK without escapeEscape option)
				{"'''AA|'BB'''", '\''},     // 1 escape char and 1 quote char
				{"'''AA||'BB'''", '\''},    // 2 escape char and 1 quote char
				{"'''AA''BB'''", '\''},     // 2 quote char (OK without escapeEscape option)
				{"'''AA|'|'BB'''", '\''},   // (1 escape char anc 1 quote char) * 2
				{"'''AA||'||'BB'''", '\''},  // (2 escape char and 1 quote char) * 2
				{"'''AA'BB'''", '|'},      // 1 quote char (OK without escapeEscape option)
				{"'''AA|'BB'''", '|'},     // 1 escape char and 1 quote char
				{"'''AA||'BB'''", '|'},    // 2 escape char and 1 quote char
				{"'''AA''BB'''", '|'},     // 2 quote char (OK without escapeEscape option)
				{"'''AA|'|'BB'''", '|'},   // (1 escape char anc 1 quote char) * 2
				{"'''AA||'||'BB'''", '|'}  // (2 escape char and 1 quote char) * 2
		};
	}

	@Test(dataProvider = "data")
	public void testEscapeEscape(String input, char escape) {
		CsvParserSettings settings = new CsvParserSettings();
		settings.setUnescapedQuoteHandling(UnescapedQuoteHandling.RAISE_ERROR);
		settings.getFormat().setQuote('\'');
		settings.getFormat().setQuoteEscape(escape);
		settings.getFormat().setCharToEscapeQuoteEscaping(escape);

		System.out.println(new CsvParser(settings).parseLine(input)[0]);
	}

```

For example, in the first input, the quote in the middle here: `AA'BB` is unescaped. In the others, such as `"'''AA||'||'BB'''"` you have escaped quotes (i.e. `'` followed by `'`) but ALSO unescaped quotes (i.e. `|` followed by `|` representing a single `|` and then the next `'` is not escaped.

Let me know if there is any particular instance of a VALID input I didn't pick up that you think should work.

**Comment by ep1804** [comment]
@jbax  

Thank you for early response. However, I can't understand your point here. I didn't prepare CSV input test cases. `AA"BB`, `AA\"BB` or `AA\\"BB` are RAW TEXTs. There's no invalid data here.

These texts should be written into CSV and recovered without change. I tried several combinations of quote/escape/escapeEscape options but I couldn't proceed without setting `quote` = `charToEscapeQuoteEscape`.

**Comment by jbax** [comment]
Can you send me the input file you are using, instead of code with strings? That's probably adding to the confusion.

**Comment by jbax** [comment]
In addition, the input you provided `AA"BB`, `AA\"BB` or `AA\\"BB` is not within quotes, so this will not be escaped by default as these values must be inside quotes.

If you want to process escape sequences in unquoted fields use `parserSettings.setEscapeUnquotedValues(true);`

**Comment by ep1804** [comment]
Here's the test code I wrote into spark source.

https://github.com/ep1804/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala#L468-L496

I'll show you java test code, when time is available. I'll try the `setEscapeUnquotedValues` option too.

**Comment by jbax** [comment]
Question, is the Java String representation of `"""AA\"BB"""` in your Spark code?

**Comment by ep1804** [comment]
Triple-quotes is Scala's syntactic sugar for write a string AS-IS. `"""AA\"BB"""` is equivalent to Java String literal `"AA\\\"BB"`.

**Comment by jbax** [comment]
Cool, so your tests are using unquoted values (i.e. there's no open and close quote around them). In this case the parser will simply read everything as-is until it finds the next delimiter. Configuring it with different quote escape settings will have no effect for the inputs you are testing.

That said, if you set `parserSettings.setEscapeUnquotedValues(true);` then the parser will try to unescape your input. If it finds anything not properly escaped it will by default try to recover and produce a sensible output. To get an error instead use `settings.setUnescapedQuoteHandling(UnescapedQuoteHandling.RAISE_ERROR);`.

I noticed you are writing and reading in your tests. So I created [this unit test](https://github.com/uniVocity/univocity-parsers/blob/master/src/test/java/com/univocity/parsers/issues/github/Github_143.java).

It produces the following output when **writing**:

```
--------------[ AA'BB ]---------
1) Escape: ':   AA''BB
2) Escape: |:   AA|'BB

--------------[ AA|'BB ]---------
1) Escape: ':   AA|''BB
2) Escape: |:   AA|||'BB

--------------[ AA||'BB ]---------
1) Escape: ':   AA||''BB
2) Escape: |:   AA|||||'BB

--------------[ AA''BB ]---------
1) Escape: ':   AA''''BB
2) Escape: |:   AA|'|'BB

--------------[ AA|'|'BB ]---------
1) Escape: ':   AA|''|''BB
2) Escape: |:   AA|||'|||'BB

--------------[ AA||'||'BB ]---------
1) Escape: ':   AA||''||''BB
2) Escape: |:   AA|||||'|||||'BB
```

All values are parsed back and are equal to the given input. No need to deal with escape-escape.

Now for **parsing** without doing anything with the escape-escape, and with  `parserSettings.setEscapeUnquotedValues(true);`:

```
--------------[ AA'BB ]---------
1) Escape: ':   AA'BB   << error
2) Escape: |:   AA'BB   << error

--------------[ AA|'BB ]---------
1) Escape: ':   AA|'BB   << error
2) Escape: |:   AA'BB

--------------[ AA||'BB ]---------
1) Escape: ':   AA||'BB   << error
2) Escape: |:   AA|'BB   << error

--------------[ AA''BB ]---------
1) Escape: ':   AA'BB
2) Escape: |:   AA''BB   << error

--------------[ AA|'|'BB ]---------
1) Escape: ':   AA|'|'BB   << error
2) Escape: |:   AA''BB

--------------[ AA||'||'BB ]---------
1) Escape: ':   AA||'||'BB   << error
2) Escape: |:   AA||'BB   << error

```

Results followed by `<< error` indicate the input has unescaped quotes and the output produced is the result of the recovery algorithm. I don't see anything too questionable about this result except for the last value produced, where I'd expect to see `AA|'|'BB`. I will have this one adjusted.

Please let me know if there is anything else these tests didn't cover or anything you don't agree with. Thanks!

**Comment by jbax** [comment]
Reopened to fix the bug parsing unescaped quote case mentioned in the previous post.

**Comment by jbax** [comment]
Update: Bug found when when writing with `settings.setQuoteEscapingEnabled(true);`. That's a bad one. Will provide a fix today.

**Comment by jbax** [comment]
Fixed all issues mentioned and released a 2.4.0-SNAPSHOT build to include these fixes.
@ep1804 can you try this one?

Thank you for reporting the issue and for the patience. The defaults in spark are different and it took me a bit to find out how you were getting the errors you reported.

**Comment by ep1804** [comment]
Thank you for your work @jbax 

I think we found the same bug from different viewpoints, and your one is more appropriate. This fix and default settings will be considered in the spark code (maybe on Monday). Happy weekend!
